capture/capture_aes.py

```diff
@@ -0,0 +1,380 @@
+    else:
+        if sample_fixed:
+            # Use fixed_key as this key.
+            key = np.asarray(key_fixed)
+        else:
+            # Generate this key from the PRNG.
+            key = bytearray(key_length)
+            for i in range(0, key_length):
+                key[i] = random.randint(0, 255)
+        # Always generate this plaintext from PRNG (including very first one).
+        plaintext = bytearray(16)
+        for i in range(0, 16):
+            plaintext[i] = random.randint(0, 255)
+        # Compute ciphertext for this key and plaintext.
+        cipher = AES.new(bytes(key), AES.MODE_ECB)
+        ciphertext = bytearray(cipher.encrypt(bytes(plaintext)))
+        # Determine if next iteration uses fixed_key.
+        sample_fixed = plaintext[0] & 0x1
+
+    return plaintext, key, ciphertext, sample_fixed
+
+
+def check_ciphertext(ot_aes, expected_last_ciphertext, ciphertext_len):
+    """ Compares the received with the generated ciphertext.
+
+    Ciphertext is read from the device and compared against the pre-computed
+    generated ciphertext. In batch mode, only the last ciphertext is compared.
+    Asserts on mismatch.
+
+    Args:
+        ot_aes: The OpenTitan AES communication interface.
+        expected_last_ciphertext: The pre-computed ciphertext.
+        ciphertext_len: The length of the ciphertext in bytes.
+    """
+    actual_last_ciphertext = ot_aes.read_ciphertext(ciphertext_len)
+    assert actual_last_ciphertext == expected_last_ciphertext[0:ciphertext_len], (
+        f"Incorrect encryption result!\n"
+        f"actual: {actual_last_ciphertext}\n"
+        f"expected: {expected_last_ciphertext}"
+    )
+
+
+def capture(scope: Scope, ot_aes: OTAES, capture_cfg: CaptureConfig, project, cwtarget: CWFPGA):
+    """ Capture power consumption during AES encryption.
+
+    Supports four different capture types:
+    * aes_random: Fixed key, random plaintext. TODO HOW ptx generated.
+    * aes_random_batch: Fixed key, random plaintext in batch mode.
+    * aes_fvsr: Fixed vs. random key.
+    * aes_fvsr_batch: Fixed vs. random key batch.
+
+    Args:
+        ot_aes: The OpenTitan AES communication interface.
+        expected_last_ciphertext: The pre-computed ciphertext.
+        ciphertext_len: The length of the ciphertext in bytes.
+    """
+    # Initial plaintext.
+    text = bytearray(capture_cfg.text_fixed)
+
+    # Load fixed key.
+    key_fixed = bytearray(capture_cfg.key_fixed)
+    key = key_fixed
+    logger.info(f"Initializing OT AES with key {binascii.b2a_hex(bytes(key))} ...")
+    ot_aes.write_key(key)
+
+    # Generate plaintexts and keys for first batch.
+    if capture_cfg.batch_mode and capture_cfg.capture_mode == "aes_fsvr_key":
+        ot_aes.write_fvsr_batch_generate(capture_cfg.num_segments.to_bytes(4, "little"))
+
+    if capture_cfg.batch_mode and capture_cfg.capture_mode == "aes_random":
+        ot_aes.write_init_text(text)
```

Comment:
Maybe do a single `if batch mode` and then nest the capture mode? Also the definition of the initial plaintext (`text` variable) could go here?